### PR TITLE
Improve openMP variables and macros to avoid unnecessary warnings

### PR DESCRIPTION
**Pull request by PaulWessel**
The _gmt_M_grd_loop_ and similar macros result in tons of warnings due to the problem with Windows not supporting OpenMP > 2.0 and therefore forces us to have signed loop variables.  Because GMT internals, including published API structures, uses unsigned variables for quantities that cannot be negative, this means we either have to add many casts or risk not noticing legitimate warnings in a sea of false positives.

This PR addresses this quandary by introducing a typedef _openmp_int_ that is set to `int` under Windows and `unsigned int` elsewhere, and we use casts to keep things quiet regardless of what _openmp_int_ is. I use openmp_int wherever those macro-loops are used, wherever we use OpenMP, and in places were we may want to add OpenMP in the future.

I have tested this branch for both signed and unsigned typedef and all warnings have gone away.  All tests that pass in master still pass in both compilations.  Before approving and merging we should see if anything funny is discovered by @joa-quim and/or the CI tests outside macOS.

### CODE REVIEW

**Comment by meghanrjones** [comment]
Do you want to run the full tests on this PR before merging or just go based off the few simple tests run on all PRs? Running the full tests can be done by uncommenting L7 in the tests workflow (https://github.com/GenericMappingTools/gmt/blob/no-warnings/.github/workflows/tests.yml). We would need to recomment that line before merging. The CI tests take a few hours to run.

File: src/gmt_grd.h
Comment (joa-quim) [inline src/gmt_grd.h:48]: This does not test for VisualStudio. MinGW has it too (no idea on what OMP it uses). To detect VisualStudio use
```
#ifdef _MSC_VER
```

Comment (PaulWessel) [reply src/gmt_grd.h:48]: OK, since MinGW uses gcc (right?) I can make this change.

Comment (joa-quim) [reply src/gmt_grd.h:48]: And the ``#if _WIN32`` would probably block the Intel compiler

Comment (PaulWessel) [reply src/gmt_grd.h]: OK, done

**Comment by meghanrjones** [comment]
The tests workflow had two new failures on linux and windows, though I was not able to reproduce these failures locally on linux:
499 - test/grdlandmask/greenwich.sh (Failed)
501 - test/grdlandmask/mask.sh (Failed)

The full test results are at https://github.com/GenericMappingTools/gmt/actions/runs/1720120943. I'll revert the tests workflow.

File: src/grdlandmask.c
Comment (meghanrjones) [inline src/grdlandmask.c:389]: Why are the changes for grdlandmask `int` rather than `openmp_int`?

Comment (PaulWessel) [reply src/grdlandmask.c:389]: I think I tried and got failed tests. It seems col_min could be negative and needs to be set:

```
col_min = irint (ceil (lon_w * HH->r_inc[GMT_X] - Grid->header->xy_off));
if (col_min < 0) col_min = 0;
```

and that would fail as _openmp_int_ on macs and linux where it would be unsigned and a huge positive value.

Comment (PaulWessel) [reply src/grdlandmask.c:389]: But I see elsewhere in that module I use MAX (0,....) to avoid this, so I will try first.

Comment (meghanrjones) [reply src/grdlandmask.c:389]: OK, I'll rerun the tests on windows to see if this fixes the test failures.

Comment (PaulWessel) [reply src/grdlandmask.c:389]: Thanks, I do see there is potential for negative values and hence I think it is safest to leave as ints.

Comment (meghanrjones) [reply src/grdlandmask.c:389]: After the latest commit I get the same failures in master and this branch on windows.

**Comment by PaulWessel** [comment]
OK, I will work some more on grdlandmask then.

**Comment by meghanrjones** [comment]
> OK, I will work some more on grdlandmask then.

Sorry, I wasn't clear. I do not think you need to do more work on grdlandmask. There are still failures on windows, but they are identical on master and this branch so unrelated to openMP implementation. I'm checking linux now and then will approve.

**Comment by PaulWessel** [comment]
OK, I misunderstood - too much multitasking and zooms at the same time.

**Comment by meghanrjones** [comment]
I ran into issues with the tests hanging - I think on [firmoviscous.sh](https://github.com/GenericMappingTools/gmt/blob/no-warnings/test/potential/firmoviscous.sh) so I am holding off on approving until I look into it more.

**Comment by PaulWessel** [comment]
I fixed a few things in **grdmask** related to OpenMP.  Now, when I compare the test results from the MP build with the non-OpenMP build I get the same 12 failures in master (related to lines).  So I think nothing is broken in the PR.

However, I learned that the **grdmask** with OpenMP looping over the few nodes inside a small search radius is much slower than non-OpenMP due to the overhead of setting up the extra threads just to compute a few points.  So either we need to decide during run time that the size of the radius is large enough to warrant doing it or we skip the openMP part.  That would mean an if-test and near-duplicate coding somewhere.  Given the speed up due to #6253 I think we should just yank the OpenMP part from the **-S** section.

**Comment by PaulWessel** [comment]
Let me know if this latest version is acceptable - no warnings on my side and I removed one OpenMP loop from grdmask (kept the one involving polygons).

**Review by meghanrjones — APPROVED** [review]

**Pull request merged**
